File: dfs/recover_bst.rb
# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

# @param {TreeNode} root
# @return {Void} Do not return anything, modify root in-place instead.
def recover_tree(root)
    @first = TreeNode.new(nil)
    @second = TreeNode.new(nil)
    @prev = TreeNode.new(-Float::INFINITY)

    dfs(root)
    @first.val, @second.val = @second.val, @first.val
end

# In-Order traversal
def dfs(node)
    return nil if !node

    dfs(node.left) # Process left

    # Process node
    @first = @prev if !@first.val && @prev.val >= root.val
    @second = node if @first.val && @prev.val >= root.val
    @prev = node

    dfs(node.right) # Process right
end

# Time: O(n)
# O(V + E) due to dfs
# => O(n + n - 1) due to tree
# => O(n)

# Space: O(n) due to Stack
# For O(1) you need Morris Threading Traversal which is hard.

File: dfs/recover_bst.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def recoverTree(self, root):
        self.firstNode = TreeNode(None)
        self.secondNode = TreeNode(None)
        self.prevNode = TreeNode(float('-inf'))
        self.traverse(root)
        self.firstNode.val, self.secondNode.val = self.secondNode.val, self.firstNode.val

    def traverse(self, root):
        if root == None:
            return        
        self.traverse(root.left)
        if self.firstNode.val == None and self.prevNode.val >= root.val:
            self.firstNode = self.prevNode        
        if self.firstNode.val != None and self.prevNode.val >= root.val:
            self.secondNode = root            
        self.prevNode = root        
        self.traverse(root.right)